### PR TITLE
Stream simple graph serialization / deserialization

### PR DESCRIPTION
**Pull request by leogrim**
@eishay @stkem

### CODE REVIEW

File: kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraph.scala [hunk 1]
```diff
@@ -47,14 +48,29 @@ case class SimpleGraph(vertices: ConcurrentMap[VertexId, MutableVertex] = TrieMa
...
 object SimpleGraph {
-  implicit val format: Format[SimpleGraph] = new Format[SimpleGraph] {
-    def writes(simpleGraph: SimpleGraph): JsValue = JsArray(simpleGraph.vertices.flatMap { case (vertexId, vertex) => Seq(VertexId.format.writes(vertexId), MutableVertex.format.writes(vertex)) }.toSeq)
-    def reads(json: JsValue): JsResult[SimpleGraph] = json.validate[JsArray].map { jsArray =>
-      val vertices = TrieMap[VertexId, MutableVertex]()
-      vertices ++= jsArray.value.sliding(2,2).map { case Seq(vertexId, vertex) =>
-        vertexId.as[VertexId] -> vertex.as[MutableVertex]
+  def write(graph: SimpleGraph, graphFile: File): Unit = {
+    val lines: Iterable[String] = graph.vertices.map { case (vertexId, vertex) => Json.stringify(
+      Json.arr(JsNumber(vertexId.id), Json.toJson(vertex))
```
Comment (eishay) [inline kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraph.scala:53]: why map and have it all in memory? could you do a foreach and dump it to file as you go (without keeping the seq of lines?)

**Comment by eishay** [comment]
how about adding a test, write/read from/to `File.createTempFile(prefix, suffix)`

File: kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraph.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraph.scala:53]: It's not a Seq but an Iterable, so I believe this is all lazy and enables to pass all the lines to FileUtils.writeLines which uses a BufferedOutputStream and a Java loop down the road. Looking at Graphite, the memory usage for persistence went way down so it looks like it's working as expected.

**Comment by leogrim** [comment]
I did add a test in a subsequent pull request, it was short so I merged it right away. It's in SimpleGraphTest.

File: kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraph.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/graph/app/com/keepit/graph/simple/SimpleGraph.scala:53]: Well at least that was my intent, and IntelliJ hinted me about it being a Traversable but I'm not sure what's underneath. What I see in Graphite seems to go that way but I'll double-check tomorrow morning and maybe use something more clearly lazy, either from an iterator or a stream (also have to check that the Java conversion doesn't instantiate anything).

**Comment by 2is10** [comment]
> how about adding a test, write/read from/to File.createTempFile(prefix, suffix)

Having a second `.write` method that takes an `OutputStream` or an [`OutputSupplier<? extends OutputStream>`](https://code.google.com/p/guava-libraries/source/browse/guava/src/com/google/common/io/OutputSupplier.java) (or Scala equivalent), to which the existing `.write` method would delegate would help keep the test fast (all in memory).

**Pull request merged**